forecasttools/idata_w_dates_to_df.py

```diff
@@ -7,80 +7,226 @@
+    setattr(idata, group, idata_group)
+    return idata
 
-    # NOTE: failure mode is if groups don't have the same
-    # start date
-
-    # convert received start date string to datetime object
-    start_date_as_dt = datetime.strptime(start_date_iso, "%Y-%m-%d")
-    # copy idata object to avoid modifying the original
-    idata_w_dates = idata_wo_dates.copy()
-    # modify indices of each selected group to dates
-    for (
-        group_name,
-        dim_name,
-    ) in group_dim_dict.items():
-        idata_group = getattr(idata_w_dates, group_name, None)
-        if idata_group is not None:
-            interval_size = idata_group.sizes[dim_name]
-            interval_dates = (
-                pl.date_range(
-                    start=start_date_as_dt,
-                    end=start_date_as_dt + pl.duration(days=interval_size - 1),
-                    interval="1d",
-                    closed="both",
-                    eager=True,
-                )
-                .to_numpy()
-                .astype("datetime64[ns]")
-            )
-            idata_group_with_dates = idata_group.assign_coords(
-                {dim_name: interval_dates}
-            )
-            setattr(
-                idata_w_dates,
-                group_name,
-                idata_group_with_dates,
+
+def add_time_coords_to_idata_dimensions(
+    idata: az.InferenceData,
+    groups: str | list[str],
+    variables: str | list[str],
+    dimensions: str | list[str],
+    start_date_iso: str | datetime,
+    time_step: timedelta,
+) -> az.InferenceData:
+    """
+    Modifies the time-based coordinates across
+    groups, variables, and dimensions. The
+    function checks if the group, variable,
+    and dimension exist, and then applies the
+    specified time-based coordinates.
+
+    Parameters
+    ----------
+    idata : az.InferenceData
+        The InferenceData object to modify.
+    groups : str | list[str]
+        A group or a list of groups to modify
+        (e.g., "posterior_predictive").
+    variables : str | list[str]
+        A variable or a list of variables
+        within the specified groups to modify.
+    dimensions : str | list[str]
+        A dimension or a list of dimensions
+        to modify for the variables.
+    start_date_iso : date | datetime
+        The start date for the time
+        coordinates as a datetime date or as
+        a datetime datetime.
+    time_step : timedelta
+        The time interval between each
+        coordinate (e.g., `timedelta(days=1)`).
+
+    Returns
+    -------
+    az.InferenceData
+        The modified InferenceData object with
+        updated time coordinates for the
+        specified groups, variables, and
+        dimensions.
+    """
+    inputs = [
+        (idata, az.InferenceData, "idata"),
+        (groups, (str, list), "groups"),
+        (variables, (str, list), "variables"),
+        (dimensions, (str, list), "dimensions"),
```

Comment:
Again, I think this can be duck-typed, especially as you're going to call `ensure_listlike` on them

Reply:
Why error if the user provides a `numpy.array` of strings, for instance?

Reply:
Your note here and in [this comment](https://github.com/CDCgov/forecasttools-py/pull/35#discussion_r1852584560) has been added to this issue #41 (see [this comment](https://github.com/CDCgov/forecasttools-py/issues/41#issuecomment-2491891359)).